File: packages/contracts-ts/src/live-run.ts
import { z } from 'zod';

import { idempotencyMetadataSchema } from './api';
import { runLoadoutSchema, runSchema } from './entities';
import { graphDeltaSchema, graphSnapshotSchema } from './graph';
import { runIdSchema, sequenceSchema, simTimestampSchema } from './primitives';
import {
  CONNECTION_HEALTH_SCHEMA_VERSION,
  LIVE_RUN_SCHEMA_VERSION,
  REALTIME_REDUCER_ACTION_SCHEMA_VERSION,
  RUN_COMMAND_RESPONSE_SCHEMA_VERSION,
  RUN_CREATE_REQUEST_SCHEMA_VERSION,
  SNAPSHOT_BOOTSTRAP_SCHEMA_VERSION,
  TIMELINE_ENTRY_SCHEMA_VERSION,
} from './versioning';

export const ConnectionHealthState = {
  CONNECTED: 'connected',
  DISCONNECTED: 'disconnected',
  RECONNECTING: 'reconnecting',
  CATCHING_UP: 'catching_up',
  GAP: 'gap',
  SNAPSHOT_RESYNC: 'snapshot_resync',
  SIMULATOR_PAUSED: 'simulator_paused',
  LOCALLY_PAUSED: 'locally_paused',
  STALE: 'stale',
} as const;

export const connectionHealthStateSchema = z.enum([
  'connected',
  'disconnected',
  'reconnecting',
  'catching_up',
  'gap',
  'snapshot_resync',
  'simulator_paused',
  'locally_paused',
  'stale',
]);

export type ConnectionHealthStateValue =
  (typeof ConnectionHealthState)[keyof typeof ConnectionHealthState];

export const timelineEntrySchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    sequence: sequenceSchema,
    eventId: z.string().min(1),
    eventType: z.string().min(1),
    label: z.string().min(1),
    timestamp: simTimestampSchema,
    status: z.string().min(1),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== TIMELINE_ENTRY_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported timeline entry schema version: ${String(value.schemaVersion)}`,
      });
    }
  });

export type TimelineEntryV1 = z.infer<typeof timelineEntrySchema>;

export const runReplicatedStateSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    runId: runIdSchema,
    lastAppliedSequence: sequenceSchema,
    runStatus: z.string().min(1),
    simTime: simTimestampSchema,
    graphRevision: z.number().int().min(0),
    timelineEntries: z.array(timelineEntrySchema),
    connectionHealth: connectionHealthStateSchema,
    isStale: z.boolean(),
    locallyPaused: z.boolean(),
    seenEventIds: z.array(z.string().min(1)).default([]),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== LIVE_RUN_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported live run schema version: ${String(value.schemaVersion)}`,
      });
    }
  });

export type RunReplicatedState = z.infer<typeof runReplicatedStateSchema>;

export const realtimeReducerActionSchema = z.discriminatedUnion('type', [
  z
    .object({
      type: z.literal('apply_graph_delta'),
      delta: graphDeltaSchema,
    })
    .strict(),
  z
    .object({
      type: z.literal('load_graph_snapshot'),
      snapshot: graphSnapshotSchema,
    })
    .strict(),
  z
    .object({
      type: z.literal('append_timeline_entry'),
      entry: timelineEntrySchema,
    })
    .strict(),
  z
    .object({
      type: z.literal('update_run_status'),
      runStatus: z.string().min(1),
      simTime: simTimestampSchema,
      sequence: sequenceSchema,
    })
    .strict(),
  z
    .object({
      type: z.literal('set_connection_health'),
      connectionHealth: connectionHealthStateSchema,
      isStale: z.boolean().optional(),
    })
    .strict(),
  z
    .object({
      type: z.literal('mark_gap'),
      expectedSequence: sequenceSchema,
      receivedSequence: sequenceSchema,
    })
    .strict(),
  z
    .object({
      type: z.literal('noop_duplicate'),
      eventId: z.string().min(1),
      sequence: sequenceSchema,
    })
    .strict(),
  z
    .object({
      type: z.literal('set_locally_paused'),
      locallyPaused: z.boolean(),
    })
    .strict(),
  z
    .object({
      type: z.literal('advance_sequence'),
      sequence: sequenceSchema,
      eventId: z.string().min(1),
    })
    .strict(),
]);

export type RealtimeReducerAction = z.infer<typeof realtimeReducerActionSchema>;

export const snapshotBootstrapPayloadSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    run: runSchema,
    graphSnapshot: graphSnapshotSchema,
    lastAppliedSequence: sequenceSchema,
    // Fog-of-war ambient pressure scalar in [0, 1]; null when absent/not yet computed.
    // Additive (schemaVersion unchanged). Scalar only — no asset/condition leak.
    threatTempo: z.number().min(0).max(1).nullable().optional(),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== SNAPSHOT_BOOTSTRAP_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported snapshot bootstrap schema version: ${String(value.schemaVersion)}`,
      });
    }
  });

export type SnapshotBootstrapPayloadV1 = z.infer<typeof snapshotBootstrapPayloadSchema>;

export const runCreateRequestSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    scenarioPackagePath: z.string().min(1),
    // Optional: omit to have the API draw a cryptographically random seed and
    // persist it (server-side RNG). A supplied seed pins the run deterministically.
    seed: z.number().int().optional(),
    runId: runIdSchema.optional(),
    // Phase 7 capability loadout chosen at launch. Additive optional field
    // (schemaVersion stays 1): when omitted the server persists the default loadout.
    loadout: runLoadoutSchema.nullable().optional(),
    // Optional one-line commander's intent (untrusted, non-authoritative operator free
    // text). Bounded to a single directive line. Additive optional field (schemaVersion
    // stays 1); persisted on the run.
    commanderIntent: z.string().max(280).nullable().optional(),
    // Reset-and-replay for pinned-seed scenarios. Run ids are derived from
    // (seed, scenarioVersionId), so a pinned seed resolves to exactly one run id for the
    // lifetime of a database: relaunching returns the run created the first time, which by
    // then is finished. When set the server destroys that run (owner-or-admin only) and
    // creates it again from scratch. Additive optional field (schemaVersion stays 1);
    // omitted keeps the resume behaviour.
    restartExisting: z.boolean().optional(),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== RUN_CREATE_REQUEST_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported run create request schema version: ${String(value.schemaVersion)}`,
      });
    }
  });

export type RunCreateRequestV1 = z.infer<typeof runCreateRequestSchema>;

export const runCommandResponseSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    run: runSchema,
    eventsEmitted: z.number().int().min(0),
    idempotency: idempotencyMetadataSchema.optional(),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== RUN_COMMAND_RESPONSE_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported run command response schema version: ${String(value.schemaVersion)}`,
      });
    }
  });

export type RunCommandResponseV1 = z.infer<typeof runCommandResponseSchema>;

export const connectionHealthSnapshotSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    runId: runIdSchema,
    connectionHealth: connectionHealthStateSchema,
    lastAppliedSequence: sequenceSchema,
    isStale: z.boolean(),
    locallyPaused: z.boolean(),
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== CONNECTION_HEALTH_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported connection health schema version: ${String(value.schemaVersion)}`,
      });
    }
  });

export type ConnectionHealthSnapshotV1 = z.infer<typeof connectionHealthSnapshotSchema>;

export const REALTIME_REDUCER_ACTION_SCHEMA_VERSION_EXPORT = REALTIME_REDUCER_ACTION_SCHEMA_VERSION;

File: packages/contracts-python/src/aegis_contracts/live_run.py
"""Phase 13 live command-centre contracts."""

from __future__ import annotations

from enum import StrEnum
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from aegis_contracts.api import IdempotencyMetadataV1
from aegis_contracts.entities import RunV1
from aegis_contracts.graph import GraphDeltaV1, GraphSnapshotV1
from aegis_contracts.versioning import (
    assert_supported_schema_version,
)


class ConnectionHealthState(StrEnum):
    CONNECTED = "connected"
    DISCONNECTED = "disconnected"
    RECONNECTING = "reconnecting"
    CATCHING_UP = "catching_up"
    GAP = "gap"
    SNAPSHOT_RESYNC = "snapshot_resync"
    SIMULATOR_PAUSED = "simulator_paused"
    LOCALLY_PAUSED = "locally_paused"
    STALE = "stale"


class TimelineEntryV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion")
    sequence: int
    event_id: str = Field(alias="eventId")
    event_type: str = Field(alias="eventType")
    label: str
    timestamp: str
    status: str

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("timeline_entry", value)
        return value


class RunReplicatedState(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion")
    run_id: str = Field(alias="runId")
    last_applied_sequence: int = Field(alias="lastAppliedSequence")
    run_status: str = Field(alias="runStatus")
    sim_time: str = Field(alias="simTime")
    graph_revision: int = Field(alias="graphRevision")
    timeline_entries: list[TimelineEntryV1] = Field(alias="timelineEntries")
    connection_health: ConnectionHealthState = Field(alias="connectionHealth")
    is_stale: bool = Field(alias="isStale")
    locally_paused: bool = Field(alias="locallyPaused")
    seen_event_ids: list[str] = Field(default_factory=list, alias="seenEventIds")

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("live_run", value)
        return value


class ApplyGraphDeltaAction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["apply_graph_delta"] = "apply_graph_delta"
    delta: GraphDeltaV1


class LoadGraphSnapshotAction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["load_graph_snapshot"] = "load_graph_snapshot"
    snapshot: GraphSnapshotV1


class AppendTimelineEntryAction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["append_timeline_entry"] = "append_timeline_entry"
    entry: TimelineEntryV1


class UpdateRunStatusAction(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    type: Literal["update_run_status"] = "update_run_status"
    run_status: str = Field(alias="runStatus")
    sim_time: str = Field(alias="simTime")
    sequence: int


class SetConnectionHealthAction(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    type: Literal["set_connection_health"] = "set_connection_health"
    connection_health: ConnectionHealthState = Field(alias="connectionHealth")
    is_stale: bool | None = Field(default=None, alias="isStale")


class MarkGapAction(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    type: Literal["mark_gap"] = "mark_gap"
    expected_sequence: int = Field(alias="expectedSequence")
    received_sequence: int = Field(alias="receivedSequence")


class NoopDuplicateAction(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    type: Literal["noop_duplicate"] = "noop_duplicate"
    event_id: str = Field(alias="eventId")
    sequence: int


class AdvanceSequenceAction(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    type: Literal["advance_sequence"] = "advance_sequence"
    sequence: int
    event_id: str = Field(alias="eventId")


_ReducerActionUnion = (
    ApplyGraphDeltaAction
    | LoadGraphSnapshotAction
    | AppendTimelineEntryAction
    | UpdateRunStatusAction
    | SetConnectionHealthAction
    | MarkGapAction
    | NoopDuplicateAction
    | AdvanceSequenceAction
)

RealtimeReducerAction = Annotated[
    _ReducerActionUnion,
    Field(discriminator="type"),
]


class SnapshotBootstrapPayloadV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion")
    run: RunV1
    graph_snapshot: GraphSnapshotV1 = Field(alias="graphSnapshot")
    last_applied_sequence: int = Field(alias="lastAppliedSequence")

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("snapshot_bootstrap", value)
        return value


class RunCreateRequestV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion")
    scenario_package_path: str = Field(alias="scenarioPackagePath")
    # Optional: when omitted the API draws a cryptographically random seed in
    # [1, 2^31-1] and persists it (server-side RNG). A supplied seed pins the run
    # deterministically (golden tests always pin their seeds). Additive change —
    # schemaVersion stays 1; existing clients that send a seed are unaffected.
    seed: int | None = None
    run_id: str | None = Field(default=None, alias="runId")

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("run_create_request", value)
        return value


class RunCommandResponseV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion")
    run: RunV1
    events_emitted: int = Field(alias="eventsEmitted")
    idempotency: IdempotencyMetadataV1 | None = None

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("run_command_response", value)
        return value


class ConnectionHealthSnapshotV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion")
    run_id: str = Field(alias="runId")
    connection_health: ConnectionHealthState = Field(alias="connectionHealth")
    last_applied_sequence: int = Field(alias="lastAppliedSequence")
    is_stale: bool = Field(alias="isStale")
    locally_paused: bool = Field(alias="locallyPaused")

    @field_validator("schema_version")
    @classmethod
    def validate_schema_version(cls, value: int) -> int:
        assert_supported_schema_version("connection_health", value)
        return value
